### PR TITLE
feat(ecs): add API support for Describe Clusters

### PR DESCRIPTION
**Pull request by paragbhingre**
Add support for getting capacity providers using Describe Cluster API.

Backend changes to support: [spinnaker/spinnaker#5400](https://github.com/spinnaker/spinnaker/issues/5400)

### CODE REVIEW

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:71]: This logic is identical. To improve maintainability, we should avoid copying and pasting duplicate code. Instead it would be beneficial to move this logic into a separate method that can be called in both places.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:50]: Might be better to rename this as getAllEcsClusterDetails (rather than description as that's a bit misleading)

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/controllers/EcsClusterController.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/controllers/EcsClusterController.java:43]: Can we keep the path to be the same as the method name? (for instance if you're using getAllEcsClustersDescription, then the path would be `/ecs/ecsClustersDescription/{account}/{region}`

**Review by piradeepk — CHANGES_REQUESTED** [review]

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (allisaurus) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:55]: these 2 conditions should be reversed, otherwise if `ecsClusters` is null then I believe `ecsClusters.size()` will throw an NPE.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (allisaurus) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:53]: we need to filter this list by region & account, don't we?

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:60]: Would it make sense to use !isEmpty rather than size > 0 (whichever is your preference). Can you do a null check here as well? 

`if (describeClusterResponse != null and !describeClusterResponse.isEmpty()) `

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:85]: What happens if you pass an empty list of clusters? Does this return an error?

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (paragbhingre) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java]: Added in the new commit.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/controllers/EcsClusterController.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/controllers/EcsClusterController.java:43]: instead of getAll you just want to call it ecsClusterDetails (see ecsClusters and getAllEcsClusters for reference.

`"/ecs/ecsClusterDetails/{account}/{region}"`

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java]: This comment still needs to be addressed.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:71]: You should check that describeClusterResponse is not null as well. Please see comment from previous revision for more details.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (paragbhingre) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:104]: In that case we will get error in failures something like :

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:83]: This logic contains duplicated code, please refactor it into separated methods.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:42]: Can you add a comment here to explain why you've set a static count of 100? (for future reference of the next person who looks at this codebase)

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:79]: Same comment as above.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:101]: You should check that describeClusterResponse is not null (since describeClusterResponse.size will throw an NPE if getClusters is null)

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:104]: In this case, it seems like the wrong approach to knowingly send an erroneous describeClusterRequest and expect a list of failures. It might make more sense to validate your inputs before you make the call in the first place (prevent unnecessary request/rate limiting, and it shouldn't be an error from the describe call if the input you provide this method is invalid)

**Review by piradeepk — CHANGES_REQUESTED** [review]

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:104]: You can just return an emptyList if clusterNames is null or empty

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (paragbhingre) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java]: resolved in the latest commit, please take a look at it and let me know if that addresses your concern.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (paragbhingre) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java]: Resolved

File: clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/EcsControllersSpec.java
Comment (piradeepk) [inline clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/EcsControllersSpec.java:105]: nit: Since you're comparing 2 values, you might want to use assertEquals here (it provides the user with a better error "expected * actual *")

File: clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/EcsControllersSpec.java
Comment (piradeepk) [inline clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/EcsControllersSpec.java:106]: nit: Since you're comparing 2 values, you might want to use assertEquals here (it provides the user with a better error "expected * actual *")

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/controllers/EcsClusterController.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/controllers/EcsClusterController.java:43]: Since this is already define this as a GET operation, typically we don't add the operation type to the path. For instance, instead of adding getEcsClusterDescriptions to the path you might want to update it to be ecsClusterDescriptions like below:

Current: ("/ecs/getEcsClusterDescriptions/{account}/{region}")

to

Updated: ("/ecs/ecsClusterDescriptions/{account}/{region}")

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:56]: Not a comment but rather curious, what inputs did you have in mind?

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:63]: It could be possible for cluster.getRegion and getAccount to be null. in this case it would lead to potential NPEs. One way to avoid this is to reverse your comparison:

For instance, you could update the

Current: `cluster.getRegion().equals(region) && cluster.getAccount().equals(account))`

to 

Updated: `region.equals(cluster.getRegion()) && account.equals(cluster.getAccount())`

In this case, if cluster.getRegion or cluster.getAccount is null, this won't thrown an NPE.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:44]: nit: Usual styling in this case would be to add a comment above the property itself.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (paragbhingre) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:56]: Good question. So, Describe Cluster API responds with ATTACHMENTS | SETTINGS | STATISTICS | TAGS only when they are included in the include[] input. e.g. if you want description of TestCluster with it's SETTINGS then you would send TestCluster inside` Clusters["TestCluster"] `and include settings like this `include["SETTINGS"].`

File: clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/EcsControllersSpec.java
Comment (piradeepk) [inline clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/EcsControllersSpec.java:107]: You'll want to reverse your arguments (assertEquals(expected, actual)). Right now your expected value is in the actual place holder.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:56]: Got it! That's awesome! Do you see a need for any of those at the moment?

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (piradeepk) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:66]: Not a blocker, but could you add some debug logging here? It would be great to be able to enable debugging and be able to determine the size of the batchClusterList (how many partitions did we split the original list into?)

**Comment by piradeepk** [comment]
Lgtm, will hold off approving until unit tests are added.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (paragbhingre) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:56]: No, we do not need them for now.

File: clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/EcsControllersSpec.java
Comment (paragbhingre) [reply clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/EcsControllersSpec.java:107]: oops, gotcha. I will do it.

**Comment by paragbhingre** [comment]
Added unit test. Please review @piradeepk @allisaurus

File: clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProviderSpec.groovy
Comment (allisaurus) [inline clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProviderSpec.groovy:132]: nit: "only"

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (allisaurus) [inline clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:56]: nit: as phrased, makes it sound like we're missing something we intend to add, when it's really that we don't need the additional info yet. Suggest amending to say "...in the future **if additional properties are needed**"

**Review by allisaurus — APPROVED** [review]
LGTM! Approved pending @piradeepk 's comments have been addressed.

File: clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/EcsControllersSpec.java
Comment (paragbhingre) [reply clouddriver-ecs/src/integration/java/com/netflix/spinnaker/clouddriver/ecs/test/EcsControllersSpec.java:96]: Right now there are no extra values that we are ignoring, but in the future if API responds with any extra value in that case our code might fail to deserialize the response. So it is already future proof.

File: clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProviderSpec.groovy
Comment (paragbhingre) [reply clouddriver-ecs/src/test/groovy/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProviderSpec.groovy:132]: Resolved.

File: clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java
Comment (paragbhingre) [reply clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsClusterProvider.java:56]: make sense.

**Review by piradeepk — APPROVED** [review]
Ship it! 💯

**Pull request merged**
